Reinstate `era` in polyfill, and fix calendar subclassing

ptomato opened the pull request:
These two tasks are in the same pull request because they would otherwise cause conflicts with rebasing, when reviewing one after the other.

(As it is, this will probably conflict with the removal of time calendars in #1188, but that shouldn't be too difficult of a rebase.)

Closes: #1046
Closes: #847
See: #300

littledan commented on spec/calendar.html:
```diff
@@ -313,189 +387,354 @@ <h1>get Temporal.Calendar.prototype.id</h1>
-        1. Throw an *Error* exception.
+        1. Let _calendar_ be the *this* value.
+        1. Perform ? RequireInternalSlot(_calendar_, [[InitializedTemporalCalendar]]).
+        1. If _calendar_.[[Identifier]] is not *"iso8601"*, throw an *Error* exception.
```
I guess this line and other similar lines should be an assertion, as it's not possible to construct a calendar which isn't iso8601.

littledan commented on polyfill/lib/calendar.mjs:
```diff
@@ -177,15 +213,15 @@ export class Calendar {
       if (ES.Type(item) === 'Object' && !('calendar' in item)) return item;
```
Aside: Didn't we decide to remove the `item = item.calendar` path?

justingrant replied:
AFAIK, no. As noted in https://github.com/tc39/proposal-temporal/issues/925#issuecomment-709544520 we decided to accept `TimeZone.from({timeZone})` and Calendar.from({calendar})` to be consistent with all other Temporal methods which accept a Temporal instance, a string, or a property bag. 

One use case that this pattern makes better is when you want to use an existing ZonedDateTime or PlainXXX as a property bag that contains the TimeZone and/or Calendar that you want to use. For example: 
```javascript
zdt1.withTimeZone(zdt2);
date1.withCalendar(date2);
instant.toZonedDateTime(zomeExistingZonedDateTime);
zdt1.toPlainDateTime().toZonedDateTime(zdt2);
```

I think there's another use case for this pattern but I'm forgetting what it is.

littledan replied:
OK, I noted in https://github.com/tc39/proposal-temporal/issues/925#issuecomment-713125950 that I disagree with this conclusion, but we can revisit this later.

littledan commented on polyfill/lib/calendar.mjs:
```diff
@@ -466,41 +400,35 @@ const gre = {
-    addEraProperties();
-  }
-
+impl['gregory'] = ObjectAssign({}, impl['iso8601'], {
```
Yasss, classic object orientation.

littledan reviewed (approved):
It's possible I missed something, but LGTM. This is what I was picturing for both era and #300.

justingrant commented:
> This makes a small change to the constructor of Temporal.TimeZone so that
it doesn't accept any arbitrary ID, only a recognized built-in one, to
bring it in line with calendars.

Per https://github.com/tc39/proposal-temporal/issues/847#issuecomment-727674433, this change would require users to choose an arbitrary and perhaps unrelated built-in ID in order to subclass a time zone. Is this really what we want?  Take a look at my concerns in the comment above-- OK to reply there.

justingrant reviewed (commented):
I admit that I don't understand the change to require a built-in ID for both calendars and time zones.  It seems like it makes custom time zones and calendars harder to implement and introduces potential error cases that didn't exist before, like `id` and `toString()` getting out of sync. What is the benefit of this change?

https://github.com/tc39/proposal-temporal/issues/847#issuecomment-727674433 has more details about these concerns.

Ms2ger commented on polyfill/lib/plaindate.mjs:
```diff
@@ -61,6 +61,10 @@ export class PlainDate {
     return GetSlot(this, CALENDAR);
   }
+  get era() {
+    if (!ES.IsTemporalDate(this)) throw new TypeError('invalid receiver');
```
The polyfill supports 402.

justingrant replied:
What's the relationship between the Temporal polyfill and a future 402 polyfill? Once the latter is available, will the Temporal polyfill remove `era` and let the 402 polyfill add it along with other planned non-ISO properties like `yearType`, `inLeapMonth`, etc. ?

justingrant replied:
Got it. So once 402 decides which custom calendar units should be available, then they'll be added to the polyfill here just like `era`?  My understanding from #1195 was that some other 402-centric code would add these extra properties to the prototype, but  I may be misunderstanding what that issue meant.

Ms2ger commented on polyfill/lib/calendar.mjs:
```diff
@@ -564,60 +492,40 @@ const jpn = {
-}
-export function GetISO8601Calendar() {
-  return GetBuiltinCalendar('iso8601');
+  return ArrayIncludes.call(BUILTIN_CALENDAR_IDS, id);
```
A 402 polyfill will update this file; a 402 engine implementation won't affect the polyfill.

ptomato replied:
Right, this array isn't observable from anywhere. The 402 calendars will be implemented by this polyfill in the future, we don't need to expose hooks for other polyfills to add calendars (and can't, if remaining compliant with the spec)

Ms2ger commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -21,7 +21,6 @@ import ToPrimitive from 'es-abstract/2020/ToPrimitive.js';
 import ToString from 'es-abstract/2020/ToString.js';
 import Type from 'es-abstract/2020/Type.js';
 
-import { GetISO8601Calendar } from './calendar.mjs';
```
This fixes #1008.

Ms2ger commented on spec/calendar.html:
```diff
@@ -399,7 +399,14 @@ <h1>Temporal.Calendar.prototype.dateFromFields ( _fields_, _options_, _construct
-        1. <mark>TODO.</mark>
+        1. Set _options_ to ? NormalizeOptionsObject(_options_).
+        1. Let _overflow_ be ? ToTemporalOverflow(_options_).
+        1. Set _fields_ to ? ToTemporalDateFields(_fields_).
```
Argument _fieldNames_ is missing

justingrant commented on polyfill/lib/calendar.mjs:
```diff
@@ -45,126 +45,162 @@ export class Calendar {
+    return impl[GetSlot(this, CALENDAR_ID)].nanosecond(time);
   }
   toString() {
     if (!ES.IsTemporalCalendar(this)) throw new TypeError('invalid receiver');
```
SGTM as long as docs for custom time zones and custom calendars are updated.  

> Worst case for serialization is that the correct ISO date (/...) is returned with an incorrect `[c=..]` annotation, right? That doesn't seem all that terrible.

It would prevent deserializing into your custom calendar but probably wouldn't crash anything, so might be an easy bug to slip into production.  But this is an advanced use case so I'm not worried much either way as long as it's documented clearly what the developer should do.

ptomato replied:
I've updated the docs (which I had forgotten, and needed to be updated anyway)

justingrant commented on polyfill/lib/calendar.mjs:
```diff
@@ -16,20 +13,23 @@ import {
-    if (!ID_REGEX.exec(id)) throw new RangeError(`invalid calendar identifier ${id}`);
-    CreateSlots(this);
     id = ES.ToString(id);
+    if (!IsBuiltinCalendar(id)) throw new RangeError(`invalid calendar identifier ${id}`);
```
The docs should probably have a sentence explaining what to do with `super()`.  Also need to document that implementing `toString()` is now mandatory when before it wasn't needed.

ptomato replied:
I've updated the docs to include this.

ptomato commented on polyfill/lib/calendar.mjs:
```diff
@@ -178,15 +213,15 @@ export class Calendar {
     }
     const stringIdent = ES.ToString(item);
-    if (IsBuiltinCalendar(stringIdent)) return GetBuiltinCalendar(stringIdent);
+    if (IsBuiltinCalendar(stringIdent)) return new Calendar(stringIdent);
```
I don't think it's possible to distinguish it based on syntax (unless we do something confusing such as require other syntax for non-builtin calendars).

ptomato commented:
I have addressed the comments and will merge this now, if you have further comments don't let that stop you from making them! I'll address them in a follow-up.
